src/iota/transfers.c

```diff
@@ -45,171 +25,299 @@ static char *int64_to_chars(int64_t value, char *chars, unsigned int num_trytes)
+    rpad_chars(extended_tag, tx->tag, NUM_HASH_TRYTES);
     chars_to_bytes(extended_tag, tag_bytes, NUM_HASH_TRYTES);
 
     bytes_add_u32_mem(tag_bytes, tag_increment);
     bytes_to_chars(tag_bytes, extended_tag, 48);
 
-    // TODO: do we need to increment both? Probably only obsoleteTag...
-    memcpy(tx->obsoleteTag, extended_tag, 27);
     memcpy(tx->tag, extended_tag, 27);
 }
 
-static void set_bundle_hash(const BUNDLE_CTX *bundle_ctx, TX_OBJECT *txs,
-                            unsigned int num_txs)
-{
-    char bundle[81];
-    bytes_to_chars(bundle_get_hash(bundle_ctx), bundle, 48);
+void clear_transaction_char_buffer(char * buffer){
+    for(int i = 0; i < 2672; i++){
+        buffer[i] = 0;
+    }
+}
+
+void iota_wallet_construct_raw_transaction_chars(char * buffer, char * bundle_hash, iota_wallet_tx_object_t * tx){
+    clear_transaction_char_buffer(buffer);
+    buffer[2673] = '\0';
+    char *c = buffer;
+
+    c = char_copy(c, tx->signatureMessageFragment, 2187);
+    c = char_copy(c, tx->address, 81);
+    c = int64_to_chars(tx->value, c, 27);
+    c = char_copy(c, tx->tag, 27);
+    c = int64_to_chars(tx->timestamp, c, 9);
+    c = int64_to_chars(tx->currentIndex, c, 9);
+    c = int64_to_chars(tx->lastIndex, c, 9);
+    c = char_copy(c, bundle_hash, 81);
+    c = char_copy(c, ZERO_HASH, 81);
+    c = char_copy(c, ZERO_HASH, 81);
+    c = char_copy(c, tx->tag, 27);
+    c = int64_to_chars((int64_t) 0, c, 9);
+    c = int64_to_chars((int64_t) 0, c, 9);
+    c = int64_to_chars((int64_t) 0, c, 9);
+    char_copy(c, tx->nonce, 27);
+}
+
+iota_wallet_tx_object_t tx_object_buffer = {};
 
-    for (unsigned int i = 0; i < num_txs; i++) {
-        memcpy(txs[i].bundle, bundle, 81);
+void clear_tx_object_buffer(void){
+    char_copy(tx_object_buffer.address, ZERO_HASH, 81);
+    char_copy(tx_object_buffer.branchTransaction, ZERO_HASH, 81);
+    char_copy(tx_object_buffer.trunkTransaction, ZERO_HASH, 81);
+    char_copy(tx_object_buffer.bundle, ZERO_HASH, 81);
+    char_copy(tx_object_buffer.nonce, ZERO_TAG, 27);
+    char_copy(tx_object_buffer.tag, ZERO_TAG, 27);
+    char_copy(tx_object_buffer.obsoleteTag, ZERO_TAG, 27);
+
+    tx_object_buffer.timestamp = 0;
+    tx_object_buffer.value = 0;
+
+
+    for(unsigned int i = 0; i < 2187; i++){
+        tx_object_buffer.signatureMessageFragment[i] = '9';
     }
 }
 
-void prepare_transfers(char *seed, uint8_t security, TX_OUTPUT *outputs,
-                       int num_outputs, TX_INPUT *inputs, int num_inputs,
-                       char transaction_chars[][2673])
-{
-    // TODO use a proper timestamp
-    const uint32_t timestamp = 0;
+void clear_tx_signature_bytes_buffer(void){
+    for(unsigned int i = 0; i < 2187; i++){
+        tx_object_buffer.signatureMessageFragment[i] = '9';
+    }
+}
+
+
+iota_wallet_tx_receiver_t tx_receiver_func;
+void iota_wallet_set_tx_receiver_func(iota_wallet_tx_receiver_t func){
+    tx_receiver_func = func;
+}
```

Comment:
Setting a global variable via an external function is not the C-way. Either access the variables directly via `extern` or, my preferred solution, just pass them as arguments to `construct_bundle`

Reply:
True. Changed that.